File: tracking/watchInfo.py
from seriesData.series import Series

class WatchInfo:
    def __init__(self, series:Series) -> None:
        self.series = series
        self.season = 1
        self.episode = 1
        self.watchLocation = ""


    def getSeries(self) -> Series:
        return self.series


    def getEpisode(self) -> int:
        self.__checkForNewSeason()
        return self.episode


    def getSeason(self) -> int:
        self.__checkForNewSeason()
        return self.season


    def getSeasonEpisodes(self) -> int:
        self.__checkForNewSeason()
        curSeasonIndex = self.season-1
        season = self.series.seasons[curSeasonIndex]
        return season.episodes

    
    def watchLocationIsWebLink(self) -> bool:
        return self.watchLocation.startswith("https://") or self.watchLocation.startswith("http://")


    def nextEpisode(self):
        curSeasonIndex = self.season-1
        if curSeasonIndex >= len(self.series.seasons):
            return
        curSeason = self.series.seasons[curSeasonIndex]
        lastSeasonIndex = len(self.series.seasons)-1
        if curSeasonIndex == lastSeasonIndex and self.episode > curSeason.episodes:
            return
        self.episode += 1
        if lastSeasonIndex > curSeasonIndex and self.episode > curSeason.episodes:
            self.episode = 1
            self.season += 1


    def unseenEpisodes(self):
        curSeasonIndex = self.season-1
        lastSeasonIndex = len(self.series.seasons)-1
        if curSeasonIndex > lastSeasonIndex:
            raise Exception("Current season index is higher than last season index")
        if curSeasonIndex < lastSeasonIndex:
            return True
        lastSeasonEpisodes = self.series.seasons[lastSeasonIndex].episodes
        if self.episode <= lastSeasonEpisodes:
            return True
        return False
        

    def __checkForNewSeason(self):
        curSeasonIndex = self.season-1
        curSeason = self.getSeries().seasons[curSeasonIndex]
        if curSeason.episodes >= self.episode:
            return
        lastSeasonIndex = len(self.getSeries().seasons)-1
        if lastSeasonIndex > curSeasonIndex:
            self.season += 1
            self.episode = 1

File: src/data/tracking/watchInfo.ts
import { Series } from "../series";
import { logger } from '../../logger';

export class WatchInfo {
    series:Series;
    season:number;
    episode:number;
    watchLocation:string;

    constructor(series:Series) {
        this.series = series;
        this.season = 1;
        this.episode = 1;
        this.watchLocation = "";
    }

    getEpisode() {
        this.#checkForNewSeason()
        return this.episode
    }

    getSeason() {
        this.#checkForNewSeason()
        return this.season
    }

    getSeriesSeasons() {
        return this.series.seasons.length;
    }

    getSeasonEpisodes() {
        return this.#getCurrentSeason().episodes;
    }

    getSeasonName() {
        return this.#getCurrentSeason().name;
    }

    watchLocationIsWebLink() {
        return this.watchLocation.startsWith("https://") || this.watchLocation.startsWith("http://")
    }

    nextEpisode() {
        let curSeasonIndex = this.season-1;
        if (curSeasonIndex >= this.series.seasons.length) return;
        let curSeason = this.series.seasons[curSeasonIndex];
        let lastSeasonIndex = this.series.seasons.length - 1;
        if (curSeasonIndex == lastSeasonIndex && this.episode > curSeason.episodes) return;
        this.episode += 1;
        if (lastSeasonIndex > curSeasonIndex && this.episode > curSeason.episodes) {
            this.episode = 1;
            this.season += 1;
        }
    }

    unseenEpisodes() {
        let curSeasonIndex = this.season - 1;
        let lastSeasonIndex = this.series.seasons.length - 1;
        if (curSeasonIndex > lastSeasonIndex) {
            logger.info("Current season index is higher than last season index");
            return false;
        }
        if (curSeasonIndex < lastSeasonIndex) return true;
        let lastSeasonEpisodes = this.series.seasons[lastSeasonIndex].episodes;
        if (this.episode <= lastSeasonEpisodes) return true;
        return false;
    }

    toPersVersion() {
        let persVer = new WatchInfoPers();
        persVer.seriesId = this.series.id;
        persVer.season = this.season;
        persVer.episode = this.episode;
        persVer.watchLocation = this.watchLocation;
        return persVer;
    }

    #checkForNewSeason() {
        let curSeasonIndex = this.season-1;
        let curSeason = this.series.seasons[curSeasonIndex];
        if (curSeason.episodes >= this.episode) return;
        let lastSeasonIndex = this.series.seasons.length-1;
        if (lastSeasonIndex > curSeasonIndex) {
            this.season += 1;
            this.episode = 1;
        }
    }

    #getCurrentSeason() {
        this.#checkForNewSeason();
        let curSeasonIndex = this.season-1;
        return this.series.seasons[curSeasonIndex];
    }
}

export class WatchInfoPers {
    seriesId:number;
    season:number;
    episode:number;
    watchLocation:string;

    constructor() {
        this.seriesId = -1;
        this.season = 1;
        this.episode = 1;
        this.watchLocation = "";
    }

    toWatchInfo(seriesForId:Series) {
        let watchInfo = new WatchInfo(seriesForId);
        watchInfo.season = this.season;
        watchInfo.episode = this.episode;
        watchInfo.watchLocation = this.watchLocation;        
        return watchInfo;
    }
}